<!DOCTYPE html>
<html lang="zh-TW">
  <head>
    <meta charset="UTF-8" />
    <meta http-equiv="X-UA-Compatible" content="IE=edge" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <link rel="icon" href="./public/favicon.ico" />
    <meta http-equiv="cache-control" content="no-cache" />
    <title>JavaScript-syntax-testing</title>
    <link rel="stylesheet" href=" https://necolas.github.io/normalize.css/8.0.1/normalize.css" />
    <link rel="stylesheet" href="./hightlight/default.min.css" />
    <link rel="stylesheet" href="./css/main.css" />
    <link rel="stylesheet" href="./css/copybutton.css" />
    <link rel="stylesheet" href="./css/hightlight.css" />
    <script src="./hightlight/hightlight.min.js"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.11/clipboard.min.js"></script>
  </head>

  <body>
    <header>
      <nav>
        <h1>
          <span id="toggle-menu"></span>
          <a href="index.html"></a>
        </h1>
      </nav>
    </header>
    <main>
      <aside>
        <nav></nav>
      </aside>
      <article>
        <h2 id="16-4">16-4 Error.cause</h2>
        <p>
          Error 實例中的
          <code>cause</code>
          數據屬性指示導致該錯誤的具體原始原因。
        </p>
        <p>
          在捕獲錯誤時，我們可能會使用更具體或更加實用的信息對錯誤進行包裝，再將其重新拋出。
          <code>cause</code>
          屬性就用於這一場景，以便仍然可以訪問原始的錯誤。
        </p>
        <table class="standard-table">
          <thead>
            <tr>
              <th class="header" colspan="2">
                Property attributes of
                <code>Error: cause</code>
              </th>
            </tr>
          </thead>
          <tbody>
            <tr>
              <td>Writable</td>
              <td>yes</td>
            </tr>
            <tr>
              <td>Enumerable</td>
              <td>no</td>
            </tr>
            <tr>
              <td>Configurable</td>
              <td>yes</td>
            </tr>
          </tbody>
        </table>
        <h3>說明</h3>
        <p>
          <code>cause</code>
          的值可以是任何類型。你不應該假設捕獲錯誤的
          <code>cause</code>
          屬性是一個
          <code>Error</code>
          ，正如在
          <code>catch</code>
          語句中綁定的變量未必一定是
          <code>Error</code>
          。下文“將結構化的數據作為 cause”示例展示了刻意不提供
          <code>Error</code>
          作為
          <code>cause</code>
          的情況。
        </p>
        <h3>範例</h3>
        <p>
          有時，捕獲錯誤並使用新的錯誤信息將其重新拋出是很有用的。在這種情況下，你應該將原始錯誤傳遞給新的
          <code>Error</code>
          的構造函數，如下所示：
        </p>
        <pre><code class="language-js">
try {
  connectToDatabase();
} catch (err) {
  throw new Error("Connecting to database failed.", { cause: err });
}
        </code></pre>
        <p>
          為開發者理解而編寫的錯誤消息可能並不適合機器解析，因為它們會受到表達方式改變（rewording），或標點符號改變的影響，破壞已有解析器的正常使用。因此，當從函數中拋出錯誤時，你可以提供結構化的數據作為錯誤原因（即
          <code>cause</code>
          字段）供機器解析，以此替代人類可讀的錯誤消息。
        </p>
        <pre><code class="language-js">
function makeRSA(p, q) {
  if (!Number.isInteger(p) || !Number.isInteger(q)) {
    throw new Error("RSA key generation requires integer inputs.", {
      cause: { code: "NonInteger", values: [p, q] },
    });
  }
  if (!areCoprime(p, q)) {
    throw new Error("RSA key generation requires two co-prime integers.", {
      cause: { code: "NonCoprime", values: [p, q] },
    });
  }
  // rsa algorithm…
}
      </code></pre>
        <p>資料來源:</p>
        <p>
          <a
            href="https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Error/cause"
          >
            https://developer.mozilla.org/.../Error/cause
          </a>
        </p>
      </article>
    </main>
  </body>
  <script type="module" src="./js/main.js"></script>
</html>
